core/trino-main/src/main/java/io/trino/execution/executor/scheduler/FairScheduler.java

```diff
@@ -297,6 +301,32 @@ long getBlockedNanos(TaskControl task)
         return task.getBlockedNanos();
     }
 
+    public String diagnostics()
+    {
+        StringBuilder builder = new StringBuilder();
+        builder.append(queue);
+
+        builder.append("Task executor: pool=%s, active=%s, queue=%s\n".formatted(
+                executor.getPoolSize(),
+                executor.getActiveCount(),
+                executor.getQueue().size()));
+
+        builder.append("Concurrency control: slots=%s, available=%s\n".formatted(
+                concurrencyControl.totalSlots(),
+                concurrencyControl.availableSlots()));
+
+        builder.append("Reservations:\n");
+        if (concurrencyControl.totalSlots() - concurrencyControl.availableSlots() == 1) {
+            builder.append("    (pending)\n");
+        }
```

Comment:
I do not understand this one - but probably I do not have to :)

Reply:
It's to deal with an oddity in the reporting. The permit is acquired before the slot is registered, so this fakes the display of a slot that is about to be registered.